frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,851 @@
+    tensorobs_p: unimplemented_impl,
+    var_p: unimplemented_impl,
+    probs_p: unimplemented_impl,
+    cond_p: unimplemented_impl,
+    while_p: unimplemented_impl,
+    for_p: unimplemented_impl,
+    grad_p: unimplemented_impl,
+    func_p: unimplemented_impl,
+    jvp_p: unimplemented_impl,
+    vjp_p: unimplemented_impl,
+    print_p: unimplemented_impl,
+}
+
+
+def interpret_impl(ctx, jaxpr):
+    """Implement a custom interpreter for Catalyst's JAXPR operands.
+    Instead of interpreting Catalyst's JAXPR operands, we will execute
+    CUDA-quantum equivalent instructions. As these operations are
+    abstractly evaluated, they will be bound by JAX. The end result
+    is that a new transform function will be traced. If they are concretely evaluated,
+    then it outputs a result instead of a trace.
+
+    Args:
+       ctx: An interpreter with the correct context
+       jaxpr: Jaxpr to be interpreted
+
+    Returns:
+       Either concrete values or a trace that corresponds to the computed values.
+    """
+
+    # Main interpreter loop.
+    # Note the absense of loops here and branches.
+    # Normally in an interpreter loop, we would have
+    # while(True):
+    for eqn in jaxpr.eqns:
+        # This is similar to direct-call threading
+        # https://www.cs.toronto.edu/~matz/dissertation/matzDissertation-latex2html/node6.html
+        INST_IMPL.get(eqn.primitive, default_impl)(ctx, eqn)
+
+    retvals = safe_map(ctx.read, jaxpr.outvars)
+    if set(retvals).issubset(ctx.measurements):
+        raise NotImplementedError(
+            "You cannot return measurements directly from a tape when compiling for cuda quantum."
+        )
+    return retvals
+
+
+class QJIT_CUDA:
+    """Class representing a just-in-time compiled hybrid quantum-classical function.
+
+    .. note::
+
+        ``QJIT_CUDA`` objects are created by the :func:`~.qjit` decorator. Please see
+        the :func:`~.qjit` documentation for more details.
+
+    Args:
+        fn (Callable): the quantum or classical function
+    """
+
+    def __init__(self, fn):
+        self.user_function = fn
+        functools.update_wrapper(self, fn)
+
+    def get_jaxpr(self, *args):
+        """Trace :func:`~.user_function`
+
+        Args:
+            *args: either the concrete values to be passed as arguments to ``fn`` or abstract values
+
+        Returns:
+            an MLIR module
+        """
+
+        with Patcher(
+            (qml.QNode, "__call__", QFunc.__call__),
+        ):
+            func = self.user_function
+            abs_axes = {}
+            static_args = None
+            # _jaxpr and _out_tree are used in Catalyst but at the moment
+            # they have no use here in CUDA.
+            # We could also pass abstract arguments here in *args
+            # the same way we do so in Catalyst.
+            # But I think that is redundant now given make_jaxpr2
+            _jaxpr, jaxpr2, _out_type2, out_tree = trace_to_jaxpr(
```

Comment:
could you choose better names for `jaxpr2` and `_out_type2`?

Reply:
See here: https://github.com/PennyLaneAI/catalyst/pull/477/files#r1490314319 (and we can continue the discussion there)